pkg/virt-handler/vm.go

```diff
@@ -3231,6 +3255,52 @@ func (d *VirtualMachineController) updateDomainFunc(old, new interface{}) {
 	}
 }
 
+func startVirtioSerialServer(ctx context.Context, vmi *v1.VirtualMachineInstance, virtShareDir string, startedServer map[string]struct{}, nodeName string) error {
+	isolation := isolation.NewSocketBasedIsolationDetector(virtShareDir)
+	if vmi.Spec.Domain.Devices.DownwardMetrics == nil {
+		return nil
+	}
+
+	if !vmi.IsRunning() {
+		return nil
+	}
+
+	domainName := fmt.Sprintf("%s/%s", vmi.GetNamespace(), vmi.GetName())
+	_, alreadyStarted := startedServer[domainName]
+
+	if alreadyStarted {
+		return nil
+	}
+
+	launcherSocketPath, err := cmdclient.FindSocketOnHost(vmi)
+	if err != nil {
+		return fmt.Errorf("failed to get the launcher socket for VMI [%s], error: %v", vmi.GetName(), err)
+	}
+
+	res, err := isolation.Detect(vmi)
+	if err != nil {
+		return fmt.Errorf("failed to detect root directory of the vmi pod for VMI [%s], error: %v", vmi.GetName(), err)
+	}
+
+	channelPath := downwardmetricspkg.ChannelSocketPathOnHost(res.Pid())
+	err = virtioserial.RunDownwardMetricsVirtioServer(ctx, nodeName, channelPath, launcherSocketPath)
+	if err != nil {
+		return fmt.Errorf("failed to start the DownwardMetrics server for VMI [%s], erro: %v", vmi.GetName(), err)
+	}
+
+	startedServer[domainName] = struct{}{}
+	return nil
+}
+
+func stopVirtioServer(vmi *v1.VirtualMachineInstance, startedServer map[string]struct{}) {
+	if vmi.Spec.Domain.Devices.DownwardMetrics == nil {
```

Comment:
Won't the goroutines from the call to `virtioserial.RunDownwardMetricsVirtioServer` in `startVirtioSerialServer` continue to run? What about a context per server that could be canceled here?